sections/7-functional-programming.md

```diff
@@ -0,0 +1,119 @@
+## 7. Functional Programming
+
+<img src="https://raw.githubusercontent.com/monifu/scala-best-practices/master/assets/scala-logo-256.png"  align="right" width="128" height="128" />
+
+These functional rules generally transcend programming language and/or platform
+barrier. Functional abstractions can make your code simpler and easier to read.
+A deep knowledge of FP is not required (but oh-boy, it is enlightening) and
+anyone should be able to utilize these to their benefit. On top of that, in comparison
+with actor frameworks, it provides unmatched levels of type-safety and encourages
+writing safer code often checked at compile-time.
+
+Scala has several ([cats](https://typelevel.org/cats/), [scalaz](https://scalaz.github.io/7/))
+great functional programming libraries which greatly enhance functional programming
+experience.
+
+### 7.1 SHOULD embrace *Functor* capabilities
+
+[*Functor*](https://typelevel.org/cats/typeclasses/functor.html) is a simple abstraction
+which enhances your context-bounded computation
+programming experience defined by `map` function. By *context* are meant data-types
+such as `Future[_]`, `Option[_]`, `Either[E, _]`, `IO[_]`, `List[_]` (and similar
+collections), and many more. As most other functional abstractions, it is expressed
+in form of a [*type-class*](https://scalac.io/typeclasses-in-scala/).
+
+Generally speaking, *Functor* can be any data type which:
+
+1) Takes yet another data type as its parameter to form a type, hence the `[_]`,
+2) satisfies these [laws](https://typelevel.org/cats/typeclasses/functor.html).
+
+Speaking of laws and a common sense, these do not exist to obfuscate anyone but
+maintain a well-behaving `map` implementation. For example:
+
+1) Can `map` function change the order of elements in a `List[_]`? No it can't.
+   It would make it useless.
+2) Is `Set[_]` a *Functor*? No because its `map` might erase duplicated values,
```

Comment:
Also, Map is interesting case. Map's .map is map over keys and values, so it's not a functor, but you could define a functor for map, if you map only for values.